src/compiler/crystal/semantic/restrictions.cr

```diff
@@ -304,6 +304,44 @@ module Crystal
     end
   end
 
+  class GenericClassType
+    def restriction_of?(other : GenericClassInstanceType, owner)
+      # ```
+      # def foo(param : Array)
+      # end
+      #
+      # def foo(param : Array(Int32))
+      # end
+      # ```
+      #
+      # Here, self is `Array`, other is `Array(Int32)`
+
+      return unless self == other.generic_type
```

Comment:
```suggestion
      return false unless self == other.generic_type
```

Reply:
It seems that there's no way that this method will return `true` anyways, right?

Reply:
Indeed, but I'd like to keep the check explicit, otherwise the explanation below (``# `SomeGeneric` is never a restriction of `SomeGeneric(X)` ``) is not :100: % accurate, as it could become ``# `Foo` is never a restriction of `Bar(X)` `` which makes no sense.

Reply:
So better change the docs instead of hardcoding redundant condition.